### ISSUE TITLE
Issue with getting input/output operations (graph.get_operation_by_name(input_name);)

### ISSUE THREAD

**Issue by bwilson92** [SciSharp/TensorFlow.NET#489]
Getting the following error when attempting to get input/output operations:

> Unhandled exception. System.Exception: Can not convert a String into a Operation.
   at Tensorflow.Graph._as_graph_element_locked(Object obj, Boolean allow_tensor, Boolean allow_operation)
   at Tensorflow.Graph.as_graph_element(Object obj, Boolean allow_tensor, Boolean allow_operation)
   at Tensorflow.Graph.get_operation_by_name(String name)
   at ArRobotVis.Program.Main(String[] args) in filepath_to_program.cs

The issue arises from:
```
  string input_name = "input_1:0";
  string output_name = "pos_out/BiasAdd:0";
  var input_operation = graph.get_operation_by_name(input_name); 
  var output_operation = graph.get_operation_by_name("pos_out/BiasAdd:0");
```

I've installed the Tensorflow binary and binding packages and have tried various different implementations documented online but cannot seem to get it to work.

Here is the full code:
```

using System;
using System.IO;
using NumSharp;
using Tensorflow;
namespace ArRobotVis
{
    class Program
    {
        static void Main(string[] args)
        {
            // Import trained model (.pb)
            var graph = new Graph();
            string dir = "filepath/to_.pb_model";
            string pbFile = "my_model.pb";
            string output_graph = (Path.Join(dir, pbFile));
            graph.Import(output_graph, "");

            // Select Input/Output layers of model
            string input_name = "input_1:0";
            string output_name = "pos_out/BiasAdd:0";
            var input_operation = graph.get_operation_by_name(input_name);
            var output_operation = graph.get_operation_by_name("pos_out/BiasAdd:0");

            /*  
            //Use model to predict input image
            var results = with<Session, NDArray>(tf.Session(graph),
                sess => sess.run(output_operation.outputs[0], 
                    new FeedItem(input_operation.outputs[0], nd)));

            results = np.squeeze(results);
            */
        }
    }
}

```

**Comment by bwilson92** [comment]
Following [this example](https://github.com/SciSharp/TensorFlow.NET/issues/402), it _appears_ to have fixed the input/output operation commands.
 
Continuing from this, when trying to import an image for prediction, following the tutorial in the [documentation](https://tensorflownet.readthedocs.io/en/latest/ImageRecognition.html), I now get this error:

> An unhandled exception of type 'System.AccessViolationException' occurred in TensorFlow.NET.dll
> Attempted to read or write protected memory. This is often an indication that other memory is corrupt.

My model outputs an image so I tried changing the data type to NDArray etc but to no avail.

**Comment by Oceania2018** [comment]
Can you try to use
```csharp
Tensor input_operation = graph.get_operation_by_name(input_name);
Tensor output_operation = graph.get_operation_by_name("pos_out/BiasAdd:0");
```
instead of 
```csharp
var input_operation = graph.get_operation_by_name(input_name);
var output_operation = graph.get_operation_by_name("pos_out/BiasAdd:0");
```

**Comment by bwilson92** [comment]
Thanks for the reply,

It is giving the same issue when using Tensor or var atm.

> Unhandled exception. System.Exception: Can not convert a String into a Operation.
   at Tensorflow.Graph._as_graph_element_locked(Object obj, Boolean allow_tensor, Boolean allow_operation)
   at Tensorflow.Graph.as_graph_element(Object obj, Boolean allow_tensor, Boolean allow_operation)
   at Tensorflow.Graph.get_operation_by_name(String name)
   at ArRobotVis.Program.Main(String[] args) in file_path/to_program.cs

This appears to work at this stage anyway (may not work for predicting):
```
            // Select Input/Output layers of model
            string input_name = "input_1";
            string output_name = "pos_out/BiasAdd";
            var input_operation = graph.OperationByName(input_name);
            var output_operation = graph.OperationByName(output_name);
```

It works for both Tensor and var.

**Comment by Oceania2018** [comment]
Seems like you messed up the graph, you imported a graph, then modified it as an input graph to load image.

**Comment by bwilson92** [comment]
I'm following [this](uhttps://tensorflownet.readthedocs.io/en/latest/ImageRecognition.html), where the only thing I changed was (for importing):
```
var graph = new Graph();
string dir = "C:/Users/Brandon/Documents/Python";
string pbFile = "my_model.pb";
string output_graph = (Path.Join(dir, pbFile));
graph.Import(output_graph, "");
```
Instead of:
```
var graph = Graph.ImportFromPB(Path.Join(dir, pbFile));
var input_operation = graph.get_operation_by_name(input_name);
var output_operation = graph.get_operation_by_name(output_name);
```
Only because when trying to use ImportFromPB() I get an error saying:
> 'Graph' does not contain a definition for Import FromPB.

Do you know any examples online that have a full walkthrough from importing a .pb model and predicting an image using an input image?  I  understand more when seeing an application.

Thanks again!

**Comment by bwilson92** [comment]
I have gone through those examples actually, I can seem to get things working when the graph is built in c# like in these examples [1](https://github.com/SciSharp/SciSharp-Stack-Examples/blob/master/src/TensorFlowNET.Examples/HelloWorld.cs) and [2](https://github.com/SciSharp/SciSharp-Stack-Examples/blob/master/src/TensorFlowNET.Examples/BasicOperations.cs), but I cannot seem to import an externally trained .pb model and use it.

Using the code below I can get the input/output operations.
```
var graph = new Graph().as_default();
string dir = "file_path/to_model";
string pbFile = "my_model.pb";
string output_graph = (Path.Join(dir, pbFile));
graph.Import(output_graph, "");

 // Select Input/Output layers of model
string input_name = "input_1";
string output_name = "pos_out/BiasAdd";
var input_operation = graph.OperationByName(input_name);
var output_operation = graph.OperationByName(output_name);
```
However I get the following error when trying to predict an input image using the model:
> System.AccessViolationException: 'Attempted to read or write protected memory. This is often an indication that other memory is corrupt.'

Here is the section of code used to import an image and predict it using the pre-trained model:
            int input_height = 299;
            int input_width = 299;
            int input_mean = 0;
            int input_std = 255;

            string file_name = "file_path/test_im.jpeg";
            var file_reader = tf.read_file(file_name, "file_reader");
            var image_reader = tf.image.decode_jpeg(file_reader, channels: 1, name: "jpeg_reader");
            var caster = tf.cast(image_reader, tf.float32);
            var dims_expander = tf.expand_dims(caster, 0);
            var resize = tf.constant(new int[] { input_height, input_width });
            var bilinear = tf.image.resize_bilinear(dims_expander, resize);
            var sub = tf.subtract(bilinear, new float[] { input_mean });
            var normalized = tf.divide(sub, new float[] { input_std });
            var nd = normalized;

            var results = tf_with<Session, NDArray>(tf.Session(graph),
                sess => sess.run(output_operation.outputs[0], 
                    new FeedItem(input_operation.outputs[0], nd)));

**Comment by Oceania2018** [comment]
Where does and come from in your last post?

**Comment by bwilson92** [comment]
nd is the imported image to be predicted by the model. It was taken from [this](https://tensorflownet.readthedocs.io/en/latest/ImageRecognition.html) example, just removed from the function for debugging.

**Comment by Oceania2018** [comment]
@bwilson92 How can I reproduce? Can you provide a full example? Attach project zip file here?

**Comment by bwilson92** [comment]
Sorry for the late reply, it was 12 pm and I had to go to bed.

I attached the image I'm trying to predict and the TensorFlow model in ".pb" format.

[MLProject.zip](https://github.com/SciSharp/TensorFlow.NET/files/4088623/MLProject.zip)

The project was too big to send but here is the full example code:

```
using System;
using System.IO;
using NumSharp;
using Tensorflow;

using static Tensorflow.Binding;

/*
using Numpy;
using K = Keras.Backend;
using Keras;
using Keras.Models;
using Keras.Layers;
using Keras.Utils;
using Keras.Optimizers;
*/

namespace ArRobotVis
{
    class Program
    {
        static void Main(string[] args)
        {
            
            // Import trained model (.pb)
            var graph = new Graph();
            string dir = "filePath/Python";
            string pbFile = "my_model.pb";
            string output_graph = (Path.Join(dir, pbFile));
            graph.Import(output_graph, "");          
            
            // Select Input/Output layers of model
            string input_name = "input_1";
            string output_name = "pos_out/BiasAdd";
            var input_operation = graph.OperationByName(input_name);
            var output_operation = graph.OperationByName(output_name);


            int input_height = 299;
            int input_width = 299;
            int input_mean = 0;
            int input_std = 255;

            string file_name = "filePath/test_im.jpeg";
            var file_reader = tf.read_file(file_name, "file_reader");
            var image_reader = tf.image.decode_jpeg(file_reader, channels: 1, name: "jpeg_reader");
            var caster = tf.cast(image_reader, tf.float32);
            var dims_expander = tf.expand_dims(caster, 0);
            var resize = tf.constant(new int[] { input_height, input_width });
            var bilinear = tf.image.resize_bilinear(dims_expander, resize);
            var sub = tf.subtract(bilinear, new float[] { input_mean });
            var normalized = tf.divide(sub, new float[] { input_std });
            var nd = normalized;

            var results = tf_with<Session, NDArray>(tf.Session(graph),
                       sess => sess.run(output_operation.outputs[0],
                       new FeedItem(input_operation.outputs[0], nd)));

            /*
                   NDArray results = tf_with<Session, NDArray>(tf.Session(graph),
                       sess => sess.run(output_operation.outputs[0],
                       new FeedItem(input_operation.outputs[0], nd)));
            */
        }

    }

}
```

**Comment by labChariot** [comment]
I was looking at a very similar error in my code and then realised that my image tensor import process had strayed from the examples by not running the imported image through the graph _before_ attempting to inference. 

I'm still not sure of the significance of this step in the image tensor import process, but I've managed to repeat the error consistently by simply omitting it and trying to work directly with the normalised Tensor (output of the tf.divide method), like you're attempting to.

Consider the code from the example you linked to above:
```
private NDArray ReadTensorFromImageFile(string file_name,
                                int input_height = 224,
                                int input_width = 224,
                                int input_mean = 117,
                                int input_std = 1)
        {
            var graph = tf.Graph().as_default();

            var file_reader = tf.read_file(file_name, "file_reader");
            var decodeJpeg = tf.image.decode_jpeg(file_reader, channels: 3, name: "DecodeJpeg");
            var cast = tf.cast(decodeJpeg, tf.float32);
            var dims_expander = tf.expand_dims(cast, 0);
            var resize = tf.constant(new int[] { input_height, input_width });
            var bilinear = tf.image.resize_bilinear(dims_expander, resize);
            var sub = tf.subtract(bilinear, new float[] { input_mean });
            var normalized = tf.divide(sub, new float[] { input_std });

            using (var sess = tf.Session(graph))
                return sess.run(normalized);
        }
```

The critical step that I was missing (and I can't see it in your import process either) is the final:
```
using (var sess = tf.Session(graph))
                return sess.run(normalized);
```

This returns an NDArray which must be embedded within a Tensor to be passed to the graph for inferencing.

Note: all my work with TensorFlow prior to this work with TensorFlow.NET used the Python API. I can't see any directly equivalent step for importing images to tensors, which is probably how I overlooked it.

If anyone can explain to me why passing the image through the graph before trying to perform an inference is so important I'd be very interested to learn.

**Comment by bwilson92** [comment]
That appears to fix my error, thanks LabChariot!

I am getting incorrect outputs, however, but I think it's due to the input image being changed during the importing process. Which means I'm using a modified input image. I'm not quite sure how to import without the modification to the input image just yet but it's definitely a step in the right direction :D.
